New package develop mode and lock file support. (#913)
nim-lang/nimble: readme.markdown

         Hint: If 'incorrect' contains source files for building 'x', rename it to 'x'. Otherwise, prevent its installation by adding `skipDirs = @["incorrect"]` to the .nimble file.
      Failure: Validation failed

On `check` command the development mode dependencies are also validated against
the lock file. The following reasons for validation failure are possible:

* The package directory is not under version control.
* The package working copy directory is not in clean state.

> * Current VCS revision is no pushed on any remote.

Reviewer: ```suggestion
* Current VCS revision is not pushed on any remote.
```

Author: Fixed.

### nimble develop

The develop command is used for putting packages in a development mode. When

> executed with a list of packages it clones their repository and if it is

Reviewer: ```suggestion
executed with a list of packages it clones their repository. If it is
```

Author: Fixed.

executed in a package directory it adds cloned packages to the special
`nimble.develop` file. This is a special file which is used for holding the
paths to development mode dependencies of the current directory package. It has
the following structure:

```json
{
    "version": "0.1.0",
    "includes": [],
    "dependencies": []
}
```

* `version` - JSON schema version
* `includes` - JSON array of paths to included files.
* `dependencies` - JSON array of paths to Nimble packages directories.

The format for included develop files is the same as the project's develop
file, but their validation works slightly different.

Validation rules:

* The included develop files must be valid.
* The packages listed in `dependencies` section must be dependencies required
by the package's `.nimble` file and to be in the required by its version range.

> Transitive dependencies are not allowed but this may be changed in the future.

Reviewer: ```suggestion
Transitive dependencies are not allowed, but this may be changed in the future.
```

Author: Fixed.

* The packages listed in the included develop files are required to be valid
**Nimble** packages, but they are not required to be valid dependencies of the

> current project. In the last case, they are simply ignored.

Reviewer: ```suggestion
current project. In the latter case, they are simply ignored.
```

Author: Fixed.

> * The develop files of the develop mode dependencies a package are being

Reviewer: ```suggestion
* The develop files or the develop mode dependencies of a package are being
```

Not sure if I understood this one correctly; my suggestion might be wrong.

Author: Here is not `or`. `Of` is correct. Additional `of` is added.

followed and processed recursively. Finally, only one common set of develop
mode dependencies is created.
* In the final set of develop mode dependencies, it is not allowed to have more

> than one packages with the same name but with different file system paths.

Reviewer: ```suggestion
than one package with the same name but with different file system paths.
```

Author: Fixed.

Just as with the ``install`` command, a package URL may also be specified
instead of a name.

> If present the validity of the package's develop file is added to the

Reviewer: ```suggestion
If present, the validity of the package's develop file is added to the
```

Author: Fixed.

requirements for validity of the package which is determined by `nimble check`
command.

The `develop` command has a list of options:

* `-p, --path path` - Specifies the path whether the packages should be cloned.
* `-c, --create [path]` - Creates an empty develop file with the name

> `nimble.develop` in the current directory or if a path is present to the given

Reviewer: ```suggestion
`nimble.develop` in the current directory or, if a path is present, to the given
```

Author: Fixed.

without having in develop mode the full dependency tree path to them. It was a
design mistake that was not allowed at the beginning. The current workaround is
to add the transitive dependency as a dependency in the project's `.nimble`
file.

### nimble lock

The `nimble lock` command will generate or update a package lock file named
`nimble.lock`. This file is used for pinning the exact versions of the
dependencies of the package. The file is intended to be committed and used by
other developers to ensure that exactly the same version of the dependencies is
used by all developers.

Currently the lock file have the structure as in the following example:

```json
{
  "version": "0.1.0",
  "packages": {
     ...
     "chronos": {
      "version": "3.0.2",
      "vcsRevision": "aab1e30a726bb47c5d3f4a75a826981836cde9e2",
      "url": "https://github.com/status-im/nim-chronos",
      "downloadMethod": "git",
      "dependencies": [
        "stew",
        "bearssl",
        "httputils",
        "unittest2"
      ],
      "checksums": {
        "sha1": "a1cdaa77995f2d1381e8f9dc129594f2fa2ee07f"
      }
    },
    ...
  }

>   }

Reviewer: ```suggestion
```

Author: Fixed.

reverse dependencies of the package in the `nimbledata.json` file. Those
packages' names also must be in the lock file.
* `checksums` - A JSON compound object containing different checksums used for
verifying that a downloaded package is exactly the same as the pinned in the
lock file package. Currently, only `sha1` checksums are supported
* `sha1` - The *sha1* checksum of the package files.

If a lock file `nimble.lock` exists, then on performing all Nimble commands
which require searching for dependencies and downloading them in the case they
are missing (like `build`, `install`, `develop`), it is read and its content is
used to download the same version of the project dependencies by using the URL,
download method and VCS revision written in it. The checksum of the downloaded
package is compared against the one written in the lock file. In the case the
two checksums are not equal then it will be printed error message and the
operation will be aborted. Reverse dependencies are added for installed locked
dependencies just like for any other package being locally installed.

### nimble sync

The `nimble sync` command will synchronize develop mode dependencies with the

> content of the lock file. If the specified in the lock file revision is not

Reviewer: ```suggestion
content of the lock file. If the revision specified in the lock file is not
```

Author: Fixed.

> found locally tries to fetch it from the configured remotes. If it is present

Reviewer: ```suggestion
found locally, it tries to fetch it from the configured remotes. If it is present
```

Author: Fixed.

> on multiple branches tries to stay on the current one and if cannot prefers

Reviewer: ```suggestion
on multiple branches, it tries to stay on the current one, and if can't, it prefers
```

Author: Fixed.

local branches rather than remote-tracking ones. If found on more than one

> branch gives the user a choice whether to switch.

Reviewer: ```suggestion
branch, it gives the user a choice whether to switch.
```

Author: Fixed.

Sync operation will also download non-develop mode dependencies versions
described in the lock file if they are not already present in the Nimble cache.

If the `-l, --list-only` option is given then the command only lists

> development mode dependencies which working copies are out of sync without

Reviewer: ```suggestion
development mode dependencies whose working copies are out of sync, without
```

Author: Fixed.

actually syncing them and without downloading missing non-develop mode
dependencies.

**Important implementation details:**

To be able to determine whether a working copy of development mode dependency
needs to be synced, locked again, or merged with or re-based on some other
branch a special sync file is kept in the VCS directory (.git or .hg) of the
current package. It keeps a record for every development mode dependency for
its current working copy revision during the last `lock`, `sync`, or `develop`
operation. The name of the file is `<package_name>.nimble.sync`.

### nimble setup

The `nimble setup` command creates a `nimble.paths` file containing file system

> paths to the dependencies. Also includes the paths file in the `config.nims`

Reviewer: ```suggestion
paths to the dependencies. It also includes the paths file in the `config.nims`
```

Author: Fixed.
